File: Manuel/개인 활동/Class 2/괄호/bracket.cpp
#include <iostream>
#include <string>
#include <stack>

using namespace std;

string is_vps(const string& ps_string)
{
    stack<char> s;
    
    for(char ch : ps_string)
    {
        if(ch == '(')
        {
            s.push(ch);
        }
        else if(ch == ')')
        {
            if(s.empty())
            {
                return "NO";
            }
            s.pop();
        }
    }

    if(s.empty())
    {
        return "YES";
    }
    else
    {
        return "NO";
    }
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int T;
    cin >> T;

    for(int i = 0; i < T; i++)
    {
        string ps_string;
        cin >> ps_string;
        cout << is_vps(ps_string) << "\n";
    }
    
    return 0;
}

File: manuel/Class 2/괄호/bracket.py
import sys

def is_vps(ps_string):
    stack = []
    for char in ps_string:
        if char == '(':
            stack.append(char)
        elif char == ')':
            if not stack:
                return "NO"
            stack.pop()
    if not stack:
        return "YES"
    else:
        return "NO"

T = int(sys.stdin.readline())

for _ in range(T):
    ps_string = sys.stdin.readline().strip()
    print(is_vps(ps_string))
